### PR TITLE
Don't hold a lock when calling a callback in ossl_namemap_doall_names

### LINKED ISSUE #14225: EVP_SIGNATURE_settable_ctx_params() hangs in pthread_rwlock_wrlock for HMAC
````
Starting program: /home/nelson/Documents/OpenSSL/openssl/demos/providers/EVP_SIGNATURE 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
EVP_SIGNATURE:
   DSA:
      gettable: algorithm-id, digest
      settable: digest, properties
   dsaEncryption:
      gettable: algorithm-id, digest
      settable: digest, properties
   rsaEncryption:
      gettable: algorithm-id, pad-mode, digest, mgf1-digest, saltlen
      settable: pad-mode, digest, properties, mgf1-digest, mgf1-properties, saltlen
   RSA:
      gettable: algorithm-id, pad-mode, digest, mgf1-digest, saltlen
      settable: pad-mode, digest, properties, mgf1-digest, mgf1-properties, saltlen
   ED25519:
      gettable: algorithm-id
   ED448:
      gettable: algorithm-id
   ECDSA:
      gettable: algorithm-id, digest-size, digest
      settable: digest-size, digest, properties, kat
   SM2:
      gettable: algorithm-id, digest-size, digest
      settable: digest-size, digest, distid
^C
Program received signal SIGINT, Interrupt.
futex_abstimed_wait (private=0, abstime=0x0, clockid=0, expected=2, futex_word=<optimized out>)
    at ../sysdeps/nptl/futex-internal.h:284
284	../sysdeps/nptl/futex-internal.h: No such file or directory.
(gdb) bt
#0  futex_abstimed_wait (private=0, abstime=0x0, clockid=0, expected=2, futex_word=<optimized out>)
    at ../sysdeps/nptl/futex-internal.h:284
#1  __pthread_rwlock_wrlock_full (abstime=0x0, clockid=0, rwlock=0x55555556e780)
    at pthread_rwlock_common.c:830
#2  __GI___pthread_rwlock_wrlock (rwlock=0x55555556e780) at pthread_rwlock_wrlock.c:27
#3  0x00007ffff7d4f9a8 in CRYPTO_THREAD_write_lock (lock=0x55555556e780) at crypto/threads_pthread.c:94
#4  0x00007ffff7d3be4a in ossl_namemap_add_names (namemap=0x55555556a990, number=0, 
    names=0x7ffff7ee3669 "BLAKE2BMAC", separator=58 ':') at crypto/core_namemap.c:280
#5  0x00007ffff7d0a305 in construct_evp_method (algodef=0x7ffff7fa09e0 <deflt_macs>, prov=0x55555555a3a0, 
    data=0x7fffffffdca0) at crypto/evp/evp_fetch.c:191
#6  0x00007ffff7d3b525 in ossl_method_construct_this (provider=0x55555555a3a0, 
    algo=0x7ffff7fa09e0 <deflt_macs>, no_store=0, cbdata=0x7fffffffdbd0) at crypto/core_fetch.c:70
#7  0x00007ffff7d3b22f in algorithm_do_this (provider=0x55555555a3a0, cbdata=0x7fffffffdb50)
    at crypto/core_algorithm.c:65
#8  0x00007ffff7d4cc14 in provider_forall_loaded (store=0x55555555a300, found_activated=0x0, 
    cb=0x7ffff7d3b100 <algorithm_do_this>, cbdata=0x7fffffffdb50) at crypto/provider_core.c:717
#9  0x00007ffff7d4c9ef in ossl_provider_forall_loaded (ctx=0x5555555592a0, 
    cb=0x7ffff7d3b100 <algorithm_do_this>, cbdata=0x7fffffffdb50) at crypto/provider_core.c:801
#10 0x00007ffff7d3b0de in ossl_algorithm_do_all (libctx=0x5555555592a0, operation_id=3, provider=0x0, 
    pre=0x7ffff7d3b400 <ossl_method_construct_precondition>, fn=0x7ffff7d3b4e0 <ossl_method_construct_this>, 
    post=0x7ffff7d3b630 <ossl_method_construct_postcondition>, data=0x7fffffffdbd0)
    at crypto/core_algorithm.c:109
#11 0x00007ffff7d3b396 in ossl_method_construct (libctx=0x5555555592a0, operation_id=3, force_store=0, 
    mcm=0x7fffffffdce8, mcm_data=0x7fffffffdca0) at crypto/core_fetch.c:124
#12 0x00007ffff7d095f0 in inner_evp_generic_fetch (libctx=0x5555555592a0, operation_id=3, name_id=169, 
    name=0x7ffff7ebb45b "HMAC", properties=0x0, new_method=0x7ffff7d19250 <evp_mac_from_dispatch>, 
    up_ref_method=0x7ffff7d195e0 <evp_mac_up_ref>, free_method=0x7ffff7d19630 <evp_mac_free>)
    at crypto/evp/evp_fetch.c:303
#13 0x00007ffff7d0928f in evp_generic_fetch (libctx=0x5555555592a0, operation_id=3, 
    name=0x7ffff7ebb45b "HMAC", properties=0x0, new_method=0x7ffff7d19250 <evp_mac_from_dispatch>, 
    up_ref_method=0x7ffff7d195e0 <evp_mac_up_ref>, free_method=0x7ffff7d19630 <evp_mac_free>)
    at crypto/evp/evp_fetch.c:349
#14 0x00007ffff7d19243 in EVP_MAC_fetch (libctx=0x5555555592a0, algorithm=0x7ffff7ebb45b "HMAC", 
    properties=0x0) at crypto/evp/mac_meth.c:155
#15 0x00007ffff7ea073c in mac_settable_ctx_params (provctx=0x55555556cc90, macname=0x7ffff7ebb45b "HMAC")
    at providers/implementations/signature/mac_legacy.c:204
#16 0x00007ffff7ea043c in mac_hmac_settable_ctx_params (provctx=0x55555556cc90)
    at providers/implementations/signature/mac_legacy.c:223
#17 0x00007ffff7d29d7c in EVP_SIGNATURE_settable_ctx_params (sig=0x555555574330) at crypto/evp/signature.c:359
#18 0x0000555555555367 in print_signature_info (name=0x555555574680 "HMAC", data=0x555555574330)
    at EVP_SIGNATURE.c:22
#19 0x00007ffff7d3b87c in do_name (namenum=0x555555574660, data=0x7fffffffdfc8) at crypto/core_namemap.c:126
#20 0x00007ffff7d39407 in doall_util_fn (lh=0x555555568ad0, use_arg=1, func=0x0, 
    func_arg=0x7ffff7d3b840 <do_name>, arg=0x7fffffffdfc8) at crypto/lhash/lhash.c:205
#21 0x00007ffff7d39479 in OPENSSL_LH_doall_arg (lh=0x555555568ad0, func=0x7ffff7d3b840 <do_name>, 
    arg=0x7fffffffdfc8) at crypto/lhash/lhash.c:220
#22 0x00007ffff7d3b82b in lh_NAMENUM_ENTRY_doall_DOALL_NAMES_DATA (lh=0x555555568ad0, 
    fn=0x7ffff7d3b840 <do_name>, arg=0x7fffffffdfc8) at crypto/core_namemap.c:129
#23 0x00007ffff7d3b7e5 in ossl_namemap_doall_names (namemap=0x55555556a990, number=169, 
    fn=0x55555555532f <print_signature_info>, data=0x555555574330) at crypto/core_namemap.c:141
#24 0x00007ffff7d09e65 in evp_names_do_all (prov=0x55555555a3a0, number=169, 
    fn=0x55555555532f <print_signature_info>, data=0x555555574330) at crypto/evp/evp_fetch.c:540
#25 0x00007ffff7d29c9e in EVP_SIGNATURE_names_do_all (signature=0x555555574330, 
    fn=0x55555555532f <print_signature_info>, data=0x555555574330) at crypto/evp/signature.c:337
#26 0x00005555555553ec in print_signature (signature=0x555555574330, arg=0x5555555592a0) at EVP_SIGNATURE.c:33
#27 0x00007ffff7d09d52 in do_one (provider=0x55555555a3a0, algo=0x7ffff7fa0fa0 <deflt_signature+144>, 
    no_store=0, vdata=0x7fffffffe280) at crypto/evp/evp_fetch.c:483
#28 0x00007ffff7d3b22f in algorithm_do_this (provider=0x55555555a3a0, cbdata=0x7fffffffe200)
    at crypto/core_algorithm.c:65
#29 0x00007ffff7d4cc14 in provider_forall_loaded (store=0x55555555a300, found_activated=0x0, 
    cb=0x7ffff7d3b100 <algorithm_do_this>, cbdata=0x7fffffffe200) at crypto/provider_core.c:717
#30 0x00007ffff7d4c9ef in ossl_provider_forall_loaded (ctx=0x5555555592a0, 
    cb=0x7ffff7d3b100 <algorithm_do_this>, cbdata=0x7fffffffe200) at crypto/provider_core.c:801
#31 0x00007ffff7d3b0de in ossl_algorithm_do_all (libctx=0x5555555592a0, operation_id=12, provider=0x0, 
    pre=0x0, fn=0x7ffff7d09cb0 <do_one>, post=0x0, data=0x7fffffffe280) at crypto/core_algorithm.c:109
#32 0x00007ffff7d09ca0 in evp_generic_do_all (libctx=0x5555555592a0, operation_id=12, 
    user_fn=0x5555555553ba <print_signature>, user_arg=0x5555555592a0, 
    new_method=0x7ffff7d293b0 <evp_signature_from_dispatch>, free_method=0x7ffff7d291c0 <EVP_SIGNATURE_free>)
    at crypto/evp/evp_fetch.c:507
#33 0x00007ffff7d29c56 in EVP_SIGNATURE_do_all_provided (libctx=0x5555555592a0, 
    fn=0x5555555553ba <print_signature>, arg=0x5555555592a0) at crypto/evp/signature.c:325
#34 0x0000555555555486 in main (argc=0, argv=0x7fffffffe428) at EVP_SIGNATURE.c:50
````

### PR DESCRIPTION
**Pull request by mattcaswell**
We don't want to hold a read lock when calling a user supplied callback.
That callback could do anything so the risk of a deadlock is high.
Instead we collect all the names first inside the read lock, and then
subsequently call the user callback outside the read lock.

Fixes #14225

### CODE REVIEW

File: doc/internal/man3/ossl_namemap_new.pod [hunk 1]
```diff
@@ -25,9 +25,9 @@ ossl_namemap_doall_names
  int ossl_namemap_name2num(const OSSL_NAMEMAP *namemap, const char *name);
  int ossl_namemap_name2num_n(const OSSL_NAMEMAP *namemap,
                              const char *name, size_t name_len);
- void ossl_namemap_doall_names(const OSSL_NAMEMAP *namemap, int number,
-                               void (*fn)(const char *name, void *data),
-                               void *data);
+ int ossl_namemap_doall_names(const OSSL_NAMEMAP *namemap, int number,
```
Comment (paulidale) [inline doc/internal/man3/ossl_namemap_new.pod:28]: I think a note to the effect of: _if this function returns error, the callback will not have been called_ would be beneficial for this one.

I.e. provide a guarantee that either zero or all names are sent to the callback function.


Doing the same for the EVP wrappers might be worth considering too.

File: crypto/core_namemap.c [hunk 2]
```diff
@@ -116,31 +116,55 @@ int ossl_namemap_empty(OSSL_NAMEMAP *namemap)
...
+     * code. This could lead to deadlocks.
+     */
     CRYPTO_THREAD_read_lock(namemap->lock);
+    num_names = lh_NAMENUM_ENTRY_num_items(namemap->namenum);
+
+    if (num_names == 0) {
+        CRYPTO_THREAD_unlock(namemap->lock);
+        return 0;
+    }
+    cbdata.names = OPENSSL_malloc(sizeof(*cbdata.names) * num_names);
```
Comment (paulidale) [inline crypto/core_namemap.c:154]: I'd have done this with a stack and a reserved size but it's more stylistic than anything.

File: crypto/core_namemap.c [hunk 3]
```diff
@@ -116,31 +116,55 @@ int ossl_namemap_empty(OSSL_NAMEMAP *namemap)
...
+    if (cbdata.names == NULL) {
+        CRYPTO_THREAD_unlock(namemap->lock);
+        return 0;
+    }
     lh_NAMENUM_ENTRY_doall_DOALL_NAMES_DATA(namemap->namenum, do_name,
                                             &cbdata);
     CRYPTO_THREAD_unlock(namemap->lock);
+
+    for (i = 0; i < cbdata.found; i++)
```
Comment (paulidale) [inline crypto/core_namemap.c:163]: Food for thought: what about passing the entire list to the callback in one go?
Often the callbacks need to process everything and have additional overhead setting up structures etc to do this.  By having the whole list, some of this complexity could be avoided (in future).

Not a change for here but something to think about.

**Review by paulidale — COMMENTED** [review]
Looks good.

File: doc/internal/man3/ossl_namemap_new.pod (hunk 1 above)
Comment (mattcaswell) [reply doc/internal/man3/ossl_namemap_new.pod:28]: Fixed

File: crypto/core_namemap.c (hunk 3 above)
Comment (mattcaswell) [reply crypto/core_namemap.c:168]: Definitely not for this PR...

**Comment by mattcaswell** [comment]
Fixup pushed addressing the feedback. Please take another look.

**Review by paulidale — APPROVED** [review]

**Comment by mattcaswell** [comment]
Fixup pushed addressing a CI failure.

Please can someone re-approve this for inclusion in the current sprint?

File: crypto/evp/evp_fetch.c [hunk 4]
```diff
@@ -530,12 +530,12 @@ int evp_is_a(OSSL_PROVIDER *prov, int number,
     return ossl_namemap_name2num(namemap, name) == number;
 }
 
-void evp_names_do_all(OSSL_PROVIDER *prov, int number,
+int evp_names_do_all(OSSL_PROVIDER *prov, int number,
                       void (*fn)(const char *name, void *data),
                       void *data)
```
Comment (t8m) [inline crypto/evp/evp_fetch.c:535]: indentation wrong here

**Review by t8m — APPROVED** [review]
Approved provided the indentation nit is fixed.

File: crypto/evp/evp_fetch.c (hunk 4 above)
Comment (mattcaswell) [reply crypto/evp/evp_fetch.c]: Fixed

**Comment by t8m** [comment]
approval holds

**Comment by paulidale** [comment]
Merged to master.

**Pull request closed without merging**